test/workers/billing_worker_test.rb

```diff
@@ -53,4 +57,62 @@ class BillingWorkerTest < ActiveSupport::TestCase
       assert BillingWorker::Callback.new.on_complete(Sidekiq::Batch::Status.new(batch.bid), callback_options.stringify_keys)
     end
   end
+
+  test 'lock_name' do
+    assert_equal 'billing::provider:123', BillingWorker.lock_name(nil, 123)
+  end
+
+  test 'enqueues checks whether lock is needed' do
+    time = Time.utc(2019, 1, 1, 8, 0)
+
+    gateway = mock(need_lock?: true)
+    PaymentGateway.expects(:find).returns(gateway)
+    BillingWorker.expects(:enqueue_for_buyer).once.with(@buyer, time, true)
+    assert BillingWorker.enqueue(@provider, time)
+
+    gateway = mock(need_lock?: false)
+    PaymentGateway.expects(:find).returns(gateway)
+    BillingWorker.expects(:enqueue_for_buyer).once.with(@buyer, time, false)
+    assert BillingWorker.enqueue(@provider, time)
+  end
+
+  test 'perform acquires a lock' do
+    time = Time.utc(2017, 12, 1, 8, 0)
+    job_options = [@buyer.id, @provider.id, time, true]
+    billing_options = [@buyer.id, { provider_account_id: @provider.id, now: time, skip_notifications: true }]
+
+    job = BillingWorker.new
+    job.expects(:lock).twice.returns(mock acquire!: true, release!: true)
+    Finance::BillingService.expects(:call!).with(*billing_options)
+    job.perform(*job_options)
+  end
+
+  test 'perform reschedules if it fails to acquire the lock' do
+    time = Time.utc(2017, 12, 1, 8, 0)
+    job_options = [@buyer.id, @provider.id, time, true]
+    billing_options = [@buyer.id, { provider_account_id: @provider.id, now: time, skip_notifications: true }]
+
+    job = BillingWorker.new
+    job.expects(:lock).once.returns(mock acquire!: false)
+    Finance::BillingService.expects(:call!).with(*billing_options).never
+    assert_raises(BillingWorker::LockError) { job.perform(*job_options) }
+  end
+
+  test 'perform ingnores the lock when not needed' do
+    time = Time.utc(2017, 12, 1, 8, 0)
+    job_options = [@buyer.id, @provider.id, time, false]
+    billing_options = [@buyer.id, { provider_account_id: @provider.id, now: time, skip_notifications: true }]
+
+    job = BillingWorker.new
+    job.expects(:lock).never
+    Finance::BillingService.expects(:call!).with(*billing_options)
+    job.perform(*job_options)
+  end
+
+  test 'lock prevents concurrent workers for the same provider' do
```

Comment:
I guess we also need a test for the concurrency to work for different providers 🤔

Reply:
For that we would need fibers. Not sure if it's worth it.